website/src/docs/hotchocolate/performance/persisted-queries.md

```diff
@@ -2,12 +2,193 @@
+
+When presented with a query hash, Hot Chocolate will now check the specified folder for a file in the following format: `{Hash}.graphql`.
+
+Example: `0c95d31ca29272475bf837f944f4e513.graphql`
+
+This file is expected to contain the query the hash was generated from.
+
+> ⚠️ Note: Do not forget to ensure that the server has access to the directory.
+
+### Redis
+
+To load persisted queries from Redis, we have to add the following package.
+
+```bash
+dotnet add package HotChocolate.PersistedQueries.Redis
+```
+
+After this we need to specify where the persisted queries are located. Using `AddReadOnlyRedisQueryStorage()` we can point to a specific Redis database in which the persisted queries are stored.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services
+        .AddGraphQLServer()
+        .AddQueryType<Query>()
+        .UsePersistedQueryPipeline()
+        .AddReadOnlyRedisQueryStorage(services =>
+            ConnectionMultiplexer.Connect("host:port").GetDatabase());
+}
+```
+
+TODO: Is the query hash the key? What is expected?
+
+## Hashing algorithms
+
+Per default Hot Chocolate uses the MD5 hashing algorithm, but we can override this default by specifying a `DocumentHashProvider`.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services
+        // choose one of the following providers
+        .AddMD5DocumentHashProvider()
+        .AddSha256DocumentHashProvider()
+        .AddSha1DocumentHashProvider()
+
+        // GraphQL server configuration
+        .AddGraphQLServer()
+        .AddQueryType<Query>()
+        .UsePersistedQueryPipeline()
+        .AddReadOnlyFileSystemQueryStorage("./persisted_queries");
+}
+```
+
+We can also configure how these hashes are encoded, by specifying a `HashFormat` as argument:
+
+```csharp
+AddSha256DocumentHashProvider(HashFormat.Hex)
+AddSha256DocumentHashProvider(HashFormat.Base64)
+```
+
+> Note: [Relay](https://relay.dev) uses the MD5 hashing algorithm - no additional Hot Chocolate configuration is needed.
+
+# Blocking regular queries
+
+We can block regular queries and only allow the execution of persisted queries, by implementing a custom request middleware.
+
+```csharp
+public class RequestMiddleware
+{
+    private readonly HotChocolate.Execution.RequestDelegate _next;
+
+    public RequestMiddleware(HotChocolate.Execution.RequestDelegate next)
+    {
+        _next = next;
+    }
+
+    public async Task InvokeAsync(IRequestContext context)
+    {
+        if (!context.IsPersistedDocument)
+        {
+            throw new GraphQLRequestException("Not a persisted document");
+        }
+
+        await _next(context);
+    }
+}
+```
+
+We need to register this middleware like the following.
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services
+        .AddGraphQLServer()
+        .UsePersistedQueryPipeline()
+        // ...
+        .UseRequest<BlockRegularQueriesMiddleware>();
```

Comment:
This is more complex than this ... you actually need to copy the full pipeline and hook into a specific place of the chain. While in your case you would still get the execution the regular request was already executed at this point.
This is why I moved this to develop so that we can introduce a new option that handles this.

Reply:
I have most of it implemented...I will revisit it later today

Reply:
Ah I see, works for me :)

Reply:
@michaelstaib what's the status of this? Should we track it as an issue?
I would like to get these docs in for the v11 guys: is there a good (albeit probably verbose) way to restrict execution to only persisted queries? If there is, I'd like to document it here and then merge it. We can then revisit the documentation, once the proposed option has been implemented in >= v12.

Reply:
@michaelstaib I extracted this as an issue so we do not loose track of it: #3943